File: libs/document-processor/src/registry/processors/composite-timeline-channel-processor.ts
import { BlueNode } from '@blue-labs/language';
import { blueIds as conversationBlueIds } from '@blue-repository/types/packages/conversation/blue-ids';

import {
  compositeTimelineChannelSchema,
  type CompositeTimelineChannel,
  type ChannelContract,
  type ChannelEventCheckpoint,
  type MarkerContract,
} from '../../model/index.js';
import { KEY_CHECKPOINT } from '../../constants/processor-contract-constants.js';
import type {
  ChannelEvaluationContext,
  ChannelMatch,
  ChannelProcessor,
  ChannelDelivery,
} from '../types.js';

const SOURCE_CHANNEL_META_KEY = 'compositeSourceChannelKey';

/**
 * Compose a checkpoint key that namespaces a child channel under its composite.
 */
export function compositeCheckpointKey(
  compositeKey: string,
  childKey: string,
): string {
  return `${compositeKey}::${childKey}`;
}

function isChannelEventCheckpoint(
  marker: MarkerContract | undefined,
): marker is ChannelEventCheckpoint {
  return (
    marker != null &&
    Object.prototype.hasOwnProperty.call(marker, 'lastEvents') &&
    Object.prototype.hasOwnProperty.call(marker, 'lastSignatures')
  );
}

export class CompositeTimelineChannelProcessor implements ChannelProcessor<CompositeTimelineChannel> {
  readonly kind = 'channel' as const;
  readonly blueIds = [
    conversationBlueIds['Conversation/Composite Timeline Channel'],
  ] as const;
  readonly schema = compositeTimelineChannelSchema;

  /**
   * Fallback match hook that delegates to {@link evaluate} for consistent behavior.
   */
  matches(
    contract: CompositeTimelineChannel,
    context: ChannelEvaluationContext,
  ): boolean | Promise<boolean> {
    return this.evaluate(contract, context).then((result) => result.matches);
  }

  /**
   * Resolve child channels and return a delivery per matching child, enriching
   * each delivered event with the source channel key.
   */
  async evaluate(
    contract: CompositeTimelineChannel,
    context: ChannelEvaluationContext,
  ): Promise<ChannelMatch> {
    const event = context.event;
    if (!event) {
      return { matches: false };
    }

    const childKeys =
      contract.channels?.filter((key) => key.trim().length > 0) ?? [];
    if (childKeys.length === 0) {
      return { matches: false };
    }

    const resolveChannel = context.resolveChannel;
    const channelProcessorFor = context.channelProcessorFor;
    if (!resolveChannel || !channelProcessorFor) {
      throw new Error(
        'Composite timeline channel evaluation requires channel resolution helpers',
      );
    }

    const eventId = context.blue.calculateBlueIdSync(event);
    const deliveries: ChannelDelivery[] = [];

    for (const childKey of childKeys) {
      const resolved = resolveChannel(childKey);
      if (!resolved) {
        throw new Error(
          `Composite timeline channel '${context.bindingKey}' references missing channel '${childKey}'`,
        );
      }

      const childProcessor = channelProcessorFor(resolved.blueId);
      if (!childProcessor) {
        throw new Error(
          `No processor registered for child channel '${childKey}' (${resolved.blueId})`,
        );
      }

      const eventClone = event.clone();
      const childContext: ChannelEvaluationContext = {
        ...context,
        event: eventClone,
        bindingKey: childKey,
      };

      const childContract = resolved.contract;
      const eventFilter = childContract.event;
      if (eventFilter && !context.blue.isTypeOfNode(eventClone, eventFilter)) {
        continue;
      }

      const matches = await childProcessor.matches(childContract, childContext);
      if (!matches) {
        continue;
      }

      const channelizedFn = childProcessor.channelize;
      const channelizedResult = channelizedFn
        ? channelizedFn.call(childProcessor, childContract, childContext)
        : undefined;

      const eventForHandlers = channelizedResult ?? eventClone.clone();
      this.enrichEvent(eventForHandlers, childKey);

      const checkpointKey = compositeCheckpointKey(
        context.bindingKey,
        childKey,
      );
      const shouldProcess = await this.shouldProcessChild({
        childProcessor,
        childContract,
        context: {
          ...childContext,
          event: event.clone(),
        },
        checkpointKey,
        markers: context.markers,
      });

      deliveries.push({
        eventNode: eventForHandlers,
        eventId,
        checkpointKey,
        shouldProcess,
      });
    }

    return {
      matches: deliveries.length > 0,
      deliveries,
    };
  }

  /**
   * Determine recency by checking whether any child delivery should process.
   */
  async isNewerEvent(
    contract: CompositeTimelineChannel,
    context: ChannelEvaluationContext,
    lastEvent: BlueNode,
  ): Promise<boolean> {
    void lastEvent;
    const result = await this.evaluate(contract, context);
    const deliveries = result.deliveries ?? [];
    if (!result.matches || deliveries.length === 0) {
      return false;
    }
    return deliveries.some((delivery) => delivery.shouldProcess !== false);
  }

  /**
   * Attach the source child channel key to event metadata for handler logic.
   */
  private enrichEvent(event: BlueNode, childKey: string): void {
    const props = event.getProperties() ?? {};
    const meta =
      props.meta instanceof BlueNode ? props.meta.clone() : new BlueNode();
    const metaProps = meta.getProperties() ?? {};
    meta.setProperties({
      ...metaProps,
      [SOURCE_CHANNEL_META_KEY]: new BlueNode().setValue(childKey),
    });
    event.setProperties({ ...props, meta });
  }

  /**
   * Apply child-level recency rules using the composite-scoped checkpoint entry.
   */
  private async shouldProcessChild(args: {
    childProcessor: ChannelProcessor<unknown>;
    childContract: ChannelContract;
    context: ChannelEvaluationContext;
    checkpointKey: string;
    markers: ReadonlyMap<string, MarkerContract>;
  }): Promise<boolean> {
    const isNewerEvent = args.childProcessor.isNewerEvent;
    if (typeof isNewerEvent !== 'function') {
      return true;
    }

    const checkpoint = this.resolveCheckpoint(args.markers);
    const lastEvent = checkpoint?.lastEvents?.[args.checkpointKey] ?? null;
    if (!lastEvent || typeof lastEvent.clone !== 'function') {
      return true;
    }

    return await isNewerEvent.call(
      args.childProcessor,
      args.childContract,
      args.context,
      lastEvent.clone(),
    );
  }

  /**
   * Resolve the checkpoint marker for the current scope, if present.
   */
  private resolveCheckpoint(
    markers: ReadonlyMap<string, MarkerContract>,
  ): ChannelEventCheckpoint | null {
    const marker = markers.get(KEY_CHECKPOINT);
    if (!isChannelEventCheckpoint(marker)) {
      return null;
    }
    return marker;
  }
}

File: references/java-sdk/src/main/java/blue/language/processor/registry/processors/CompositeTimelineChannelProcessor.java
package blue.language.processor.registry.processors;

import blue.language.model.Node;
import blue.language.processor.ChannelEvaluationContext;
import blue.language.processor.ChannelProcessor;
import blue.language.processor.ChannelProcessorEvaluation;
import blue.language.processor.ContractBundle;
import blue.language.processor.model.ChannelContract;
import blue.language.processor.model.ChannelEventCheckpoint;
import blue.language.processor.model.CompositeTimelineChannel;
import blue.language.processor.model.MarkerContract;
import blue.language.processor.util.ProcessorContractConstants;

import java.util.ArrayList;
import java.util.List;

public class CompositeTimelineChannelProcessor implements ChannelProcessor<CompositeTimelineChannel> {

    public static String compositeCheckpointKey(String compositeKey, String childKey) {
        return compositeKey + "::" + childKey;
    }

    @Override
    public Class<CompositeTimelineChannel> contractType() {
        return CompositeTimelineChannel.class;
    }

    @Override
    public ChannelProcessorEvaluation evaluate(CompositeTimelineChannel contract, ChannelEvaluationContext context) {
        Node event = context.event();
        if (event == null || contract.getChannels().isEmpty()) {
            return ChannelProcessorEvaluation.noMatch();
        }

        List<ChannelProcessorEvaluation.ChannelDelivery> deliveries = new ArrayList<>();
        String rootEventId = TimelineEventSupport.eventId(event);

        for (String childKey : contract.getChannels()) {
            ContractBundle.ChannelBinding childBinding = context.resolveChannel(childKey);
            if (childBinding == null) {
                throw new IllegalStateException("Composite timeline channel '" + context.bindingKey()
                        + "' references missing channel '" + childKey + "'");
            }
            ChannelContract childContract = childBinding.contract();
            @SuppressWarnings("unchecked")
            ChannelProcessor<ChannelContract> childProcessor =
                    (ChannelProcessor<ChannelContract>) context.channelProcessorFor(childContract);
            if (childProcessor == null) {
                throw new IllegalStateException("No processor registered for composite child channel '" + childKey + "'");
            }

            Node childEvent = event.clone();
            ChannelEvaluationContext childContext = new ChannelEvaluationContext(
                    context.scopePath(),
                    childBinding.key(),
                    childEvent,
                    context.eventObject(),
                    context.markers(),
                    context.bundle(),
                    context.registry());

            ChannelProcessorEvaluation childEvaluation = childProcessor.evaluate(childContract, childContext);
            if (childEvaluation == null || !childEvaluation.matches()) {
                continue;
            }

            Node eventForHandlers = childEvaluation.eventNode() != null ? childEvaluation.eventNode() : childContext.event();
            if (eventForHandlers == null) {
                continue;
            }
            enrichWithCompositeSource(eventForHandlers, childBinding.key());
            String checkpointKey = compositeCheckpointKey(context.bindingKey(), childBinding.key());
            Boolean shouldProcess = shouldProcessChild(childProcessor, childContract, childContext, checkpointKey, context);
            deliveries.add(new ChannelProcessorEvaluation.ChannelDelivery(
                    eventForHandlers,
                    childEvaluation.eventId() != null ? childEvaluation.eventId() : rootEventId,
                    checkpointKey,
                    shouldProcess));
        }

        if (deliveries.isEmpty()) {
            return ChannelProcessorEvaluation.noMatch();
        }
        return ChannelProcessorEvaluation.matchedDeliveries(deliveries);
    }

    @Override
    public boolean matches(CompositeTimelineChannel contract, ChannelEvaluationContext context) {
        return evaluate(contract, context).matches();
    }

    @Override
    public boolean isNewerEvent(CompositeTimelineChannel contract, ChannelEvaluationContext context, Node lastEvent) {
        ChannelProcessorEvaluation evaluation = evaluate(contract, context);
        if (evaluation == null || !evaluation.matches()) {
            return false;
        }
        for (ChannelProcessorEvaluation.ChannelDelivery delivery : evaluation.deliveries()) {
            if (!Boolean.FALSE.equals(delivery.shouldProcess())) {
                return true;
            }
        }
        return false;
    }

    private void enrichWithCompositeSource(Node event, String childKey) {
        if (event == null || childKey == null) {
            return;
        }
        Node meta = event.getProperties() != null ? event.getProperties().get("meta") : null;
        if (meta == null) {
            meta = new Node();
            event.properties("meta", meta);
        }
        meta.properties("compositeSourceChannelKey", new Node().value(childKey));
    }

    private Boolean shouldProcessChild(ChannelProcessor<ChannelContract> childProcessor,
                                       ChannelContract childContract,
                                       ChannelEvaluationContext childContext,
                                       String checkpointKey,
                                       ChannelEvaluationContext compositeContext) {
        MarkerContract marker = compositeContext.markers().get(ProcessorContractConstants.KEY_CHECKPOINT);
        if (!(marker instanceof ChannelEventCheckpoint)) {
            return null;
        }
        Node lastEvent = ((ChannelEventCheckpoint) marker).lastEvent(checkpointKey);
        if (lastEvent == null) {
            return null;
        }
        return childProcessor.isNewerEvent(childContract, childContext, lastEvent);
    }
}
